### ISSUE TITLE
[Feature Request]  Change "lock battery" to "lock_battery" and remove "%" character on its value

### ISSUE THREAD

**Issue by rugene76** [JoshuaMulliken/ha-wyzeapi#333]
**Is your feature request related to a problem? Please describe.**
I ran into this issue when creating an automation in node-red to scan thru wyze lock battery levels looking for low levels and, if low, notifying myself with an email.

The parameter "lock battery" defined in lock.py, line 132 causes difficulty due to the space in the name when used in an automation to scan through battery levels. Also, appending the "%" character to it value complicates code that tests for battery level.
A similar situation also applies to "keypad battery"

**Describe the solution you'd like**
Requested change: 
1. change "lock battery" to "lock_battery" 
2. change "keypad battery" to "keypad_battery"
3. remove the "%" character from the value of each parameter

As far as I can tell from my testing, the following lines need to be changed in the lock.py file:
change line 132 from:
            dev_info["lock battery"] = str(self._lock.raw_dict.get("power")) + "%"
to:
            dev_info["lock_battery"] = str(self._lock.raw_dict.get("power"))

and change line 136 from:
           dev_info["keypad battery"] = str(self._lock.raw_dict.get("keypad", {}).get("power")) + "%"
to:
           dev_info["keypad_battery"] = str(self._lock.raw_dict.get("keypad", {}).get("power"))

Once the change is complete, the parameters will be in a more usable format to create automations with.

thanks
Gene

**Comment by ninadpchaudhari** [comment]
Just my 2 cents: 
- Changing from "Lock Battery" to "lock_battery" might break some existing automatons!
- Haven't used Node_red, but doesn't `battery level` automatically resolve to `battery_level` ? Checkout the developer tools for battery level in IPhone and the Lock! They look similar, I don't see how lock's battery sensor is problematic, but an IPhone's battery isn't. Iphone and Android battery values are fetched from the official HA App. So I believe that should the the gold standard to model any battery data to. 
- The '%' seems to be not creating any issue for my HA Instance in automatons. Moreover the lock's batter percentage is accurately depicted in the History tab too.

Find the attachments for both the sensors: Lock's battery & IPhone's battery

Basically, I am a noobie to this particular project, but not sure if this feature is required as the values look consistent with other reference sensor values. 
![Screenshot_20220201_133643](https://user-images.githubusercontent.com/1305825/152030739-73e93ff1-57a6-40c6-852a-3adefd3f87b4.png)
![Screenshot_20220201_133715](https://user-images.githubusercontent.com/1305825/152030741-25719fc6-e53f-4d15-a7b0-c99b08facc30.png)

**Comment by yoinx** [comment]
@rugene76 @ninadpchaudhari  makes some good points. I personally don't use many automations in HomeAssistant. I don't think that removing the % is a terrible idea. 

I do think that Home Assistant may be stripping the non-numeric characters in the developer tools though for the numbers. So it may be giving a false sense that there isn't a problem. I *do* believe that he is right though that the space in the name should be irrelevant as I believe that HomeAssistant automatically converts them to underscores for the entity anyway, no?

Do you have an example of an automation that is negatively affected with the way it currently is vice how that automation would/should look with the changes?

**Comment by rugene76** [comment]
Thanks for the comments
In regards to the sensor.xxxx_lock_battery entity: 
To be clear, my request is not about this object. This entity is as it should be and requires no changes. In fact, its in the syntactical guideline for what I've requested for the lock attributes. My request is only regarding the **lock** object and it attributes.

Regarding existing automations in HA and syntax for the lock object: 
Existing automations should be fine with adding a "_" into the "lock.attribute,lock battery" name since HA already handles the missing character and the "%". This "feature" is thanks to the coders that anticipated this situation in HA automations. Other automation platforms such as Node red do not, and should not, scrub a variable name except if explicitly requested. Otherwyze, code syntax rules get broken.

By way of background:
In node red, the node "get entities" can be used to find all entities with a common attribute, e.g. lock_battery, with no coding at all. It then passes out an object array for the next node to, as in my case, create an email. Very easy to do but requires adherence to java syntac rules. I use this capability to find all batteries in my environment, check their battery level and then warn if the level is low. 

As an aside, I'd suggest taking a look at node red - its very powerful and widely used in industry. I found the HA automation engine too limited and cumbersome and now use node red extensively to build automations in HA with ESPHome and other native devices.

regards,
Gene

**Comment by yoinx** [comment]
Thanks, that makes sense for why it would be causing issues down the chain. Once I get time I'll try to make the changes.

**Comment by yoinx** [comment]
@rugene76 I was thinking about this a bit more. I really don't use home assistant much, so I'm still trying to process this idea without causing any other oddities.

Rather than pulling the text display value from the extra state attributes on the device directly... Why aren't you able to poll the sensor entity for the battery? You mention it, so I'm sure that you know of it.

The issue that I'm seeing with removing the % from the extra state attributes is that it would make the display in the lock info look odd as there is no other way to indicate that it's a percentage on that screen. The sensor devices already handle this in the ui, and provide the raw value. This makes them seem like the logical place to pull the raw value from.

**Comment by yoinx** [comment]
![Screenshot_20220210-183259_Home Assistant](https://user-images.githubusercontent.com/1645503/153514817-bd091e5a-f1a3-4a17-b9e1-eb8a371c09a2.jpg)

**Comment by rugene76** [comment]
Having the proper battery identifier with the lock object allows simple recursion across all lock objects that checks battery level. The sensor object does not allow for that. 

Preformatting the battery level with "%" predetermines its usefulness except for display purposes. For example, filter operations require a messy unwinding of the parameter's formatting to be usable in numeric comparisons.

**Comment by yoinx** [comment]
@rugene76 thanks for the reply and sorry to keep going back and forth so much on something so trivial. However, the reason that I am asking is that the existence of these values seems like a Home Assitant anti-pattern. So, if we ever clean up the code-base and develop the test units and what not to actually try and get this integrated into Home Assistant's official addons (not saying this will ever happen)... We would have to fix some of the things like this and it will break anything that people are doing that makes use of these anti-patterns.

This specifically was only added because of my lack of understanding of how to link entities to objects. This is why I am of the impression that they should only really be here for display purposes. 

From a Home Assistant Device/Entity relationship perspective... The only things that should even be in this entity are "available" and lock cylinder associated states (locked/unlocked, locking/unlocking, jammed). The rest of these values should really be in other entities attached to the device. 

IE:
Lock Device
- Manufacturer
- Model Number
- Name
-- Lock Entity (locked/unlocked)
-- Binary Sensor Entity (open/closed)
-- Battery Sensor Entity (Lock Battery)
-- Battery Sensor Entity (Keypad Battery, if it exists)
-- Binary Sensor Entity (Door Open/Closed)

I say all that simply to explain that obtaining the information from the extra_state_attributes() field may break in the future if we start cleaning things up. So it's a simple change, and I'll push it up this weekend sometime. However, it feels like the wrong place to be pulling it from, and we shouldn't be including it here, to begin with.

I don't know how hard drilling down from the device to the entities is in NodeRed (I don't use it, and barely use Home Assistant)

**Comment by yoinx** [comment]
@rugene76  looks like this should be in 0.1.5-alpha.2

**Comment by rugene76** [comment]
thanks! much appreciated!

**Comment by JoshuaMulliken** [comment]
This is in stable!

**Comment by rugene76** [comment]
looks like the lock_battery change never got it.

Look at lock.py, line 132 of the most recent build.
 "lock battery" should be "lock_battery" (i.e., with an underscore instead of a space.)

thanks

**Comment by JoshuaMulliken** [comment]
@rugene76 can you create a PR?

**Comment by ninadpchaudhari** [comment]
Also added keypad battery on the next line.  #370

**Comment by JoshuaMulliken** [comment]
@yoinx made a good point on the PR:

> Bottom line... People should be polling the proper entities for these values, not this attributes field. These battery percentages shouldn't even actually exist here as I explained in the linked conversation.

What does everyone think about just removing the battery info from the device attributes?

Ping:
- @ninadpchaudhari 
- @rugene76

**Comment by rugene76** [comment]
I see no harm in leaving the attribute but having an entity for the lock battery probably fits better into a general framework for addressing diagnostic aspects of a device (e.g., as this integration already does.). I'd suggest putting in the "underscore" and calling it a day. There are more interesting things to spend coding time on.

**Comment by JoshuaMulliken** [comment]
Sounds good! I will get it merged and once @brg468 has a fix for the color bulb issue I will publish a new release.

**Comment by ninadpchaudhari** [comment]
> I see no harm in leaving the attribute but having an entity for the lock battery probably fits better into a general framework for addressing diagnostic aspects of a device (e.g., as this integration already does.). I'd suggest putting in the "underscore" and calling it a day. There are more interesting things to spend coding time on.

1. I loved that response. We need to focus on getting more people in! One underscore does no damage! No point in waiting time on trivial things lol

2. But, I just recently read this [developer documentation about extra attributes! ](https://developers.home-assistant.io/docs/core/entity/#:~:text=changeable.%20Learn%20more.-,WARNING,-Entities%20that%20generate )
>Entities that generate a significant amount of state changes can quickly increase the size of the database when the extra_state_attributes also change frequently. Minimize the number of extra_state_attributes for these entities by removing non-critical attributes or creating additional sensor entities.

3. With this information in mind, I think that having that battery percentage there as a state attribute is a bug! 
4. Can you guide me to a bunch of other devices that share this pattern of having battery percentage in `state_attributes`?
5. Do other professional locks/devices do this? 
6. I understand this conversation might seem trivial, but having knowledge of patterns might be helpful for some one in the long run :)
7. Again, not my project, @JoshuaMulliken  makes the final calls lol , just sharing my 2 cents and asking questions if anyone has answers :)

**Comment by JoshuaMulliken** [comment]
@ninadpchaudhari I am happy to accept a PR that removes those attributes (as long as it is identified as problematic). Do you want to start a new issue to track the potential database problem?

If this is an antipattern (or just tech debt) then we need to track it and resolve it.

**Comment by JoshuaMulliken** [comment]
@brg468 can you create a new issue to gather feedback on removing the unnecessary values? I am currently leaning toward dropping them (I don't use them) but I definitely want to gather more feedback.
